### ISSUE TITLE
how to create a complex router?

### ISSUE THREAD

**Issue by pyoner** [iann838/workery#4]
I didn't find any information in the documentation about how to create a complex router.
I want to split my code into separate files.

**Comment by pyoner** [comment]
I've organized my code like this:
```typescript
// the v0.ts
import { App } from 'workery';

export function use<E>(app: App<E>) { 
// my code here
// app.get('/', {})
}
```
```typescript
// the main file
import { App } from 'workery';

import { use } from './v0';

const app = new App({});
use(app);

export default app;

```

**Comment by iann838** [comment]
You mean as in:

```ts
// a.ts
export const aApp = new App<Env>({})
aApp.get("/", ...)
```

```ts
// index.ts
import { aApp } from "./a.ts"
const app = new App<Env>({})
aApp.extend("/a", aApp)
```

This feature is currently on my local files not yet committed, I can write some tests in a moment and push it later.

**Comment by iann838** [comment]
Some new features were added as well, finalizing tests and docs, I will push v1.1.0 tomorrow instead.

**Comment by pyoner** [comment]
@iann838 wow great!
thanks!

**Comment by iann838** [comment]
@pyoner 
I have added an alpha version to npm for personal testing: `workery@1.1.0` (with `--tag alpha`). Documentation isn't finished yet, will probably need some weekend time for it. However, if you want to use it now here is what you asked for previously:

```ts
import { Include } from "worker"

const subapp = new Include<env>({})

subapp.get("/hello-world", {
    parameters: {},
    handle: () => "Hello World!",
})
```

```ts
app.include("/subpath", subapp)
```

There is only one breaking change between 1.0 and 1.1: Init of `App` now instead of `basePath` has `rootPath`, this change was needed to differentiate the different behavior on the docs.

The API is finalized and is very unlikely to change from here until release candidate next week, so you should be safe to use it.

Although if you are not in a rush, I would suggest to wait for the new docs to come out first.

**Comment by pyoner** [comment]
@iann838 
I don't like how the new sub-routing API looks.
I prefer [Hono](https://hono.dev/docs/api/routing#grouping) sub-routing.
```typescript
const book = new Hono()

book.get('/', (c) => c.text('List Books')) // GET /book
book.get('/:id', (c) => {
  // GET /book/:id
  const id = c.req.param('id')
  return c.text('Get Book: ' + id)
})
book.post('/', (c) => c.text('Create Book')) // POST /book

const app = new Hono()
app.route('/book', book)
```

What do you think about this sub-routing API?

```typescript
// subapp.ts
import { App } from "workery"

const subapp = new App<env>({})

subapp.get("/hello-world", {
    parameters: {},
    handle: () => "Hello World!",
})
```

```typescript
// main.ts
app.route('/sub-path', subapp)
```

**Comment by iann838** [comment]
@pyoner The issue here is that `App` accepts many params that are only used for the main app. The `Include` is a parent class of `App` with only router capabilities. I personally felt conflicted by using `App` for sub-routing, with most params being unused. Any ideas, this can still be modified as there is no release candidate.

**Comment by iann838** [comment]
On closer inspection, I guess using `App` can work, but it simply ignores all those init configs if included under another app. Now the issue becomes how does middleware gets resolved when a sub-app has it.

**Comment by iann838** [comment]
@pyoner I can do `SubApp`
```ts
// subapp.ts
import { SubApp } from "workery"

const subapp = new SubApp<Env>({})

subapp.get("/hello-world", {
    parameters: {},
    handle: () => "Hello World!",
})
```
It cannot use `route` for adding sub app because it is already used for dynamic HTTP methods, so `include` feels the closest.

**Comment by iann838** [comment]
Published `1.1.0-alpha.2`
```ts
import { Router } from "workery"

const subapp = new Router<Env>({})

subapp.get("/hello-world", {
    parameters: {},
    handle: () => "Hello World!",
})
```
```ts
app.include("/subpath", subapp)
```

**Comment by pyoner** [comment]
> Published `1.1.0-alpha.2`
> 
> ```ts
> import { Router } from "workery"
> 
> const subapp = new Router<Env>({})
> 
> subapp.get("/hello-world", {
>     parameters: {},
>     handle: () => "Hello World!",
> })
> ```
> 
> ```ts
> app.include("/subpath", subapp)
> ```

This looks good!
I'll try the alpha version.

@iann838 thank you for your work!

**Comment by iann838** [comment]
`1.1.1` has been published, a complete revamp of docs, and the API remains intact as discussed here, closing.
